internal/dag/cache.go

```diff
@@ -181,42 +198,51 @@ func (kc *KubernetesCache) Insert(obj interface{}) bool {
 					kc.gatewayclass = gatewayClass
 				}
 
-				return true
+				return true, 1
 			// Otherwise, take whatever we're given.
 			default:
 				kc.gateway = obj
-				return true
+				return true, 1
 			}
+
 		case *gatewayapi_v1beta1.HTTPRoute:
 			kc.httproutes[k8s.NamespacedNameOf(obj)] = obj
-			return kc.routeTriggersRebuild(obj.Spec.ParentRefs)
+			return kc.routeTriggersRebuild(obj.Spec.ParentRefs), len(kc.httproutes)
+
 		case *gatewayapi_v1alpha2.TLSRoute:
 			kc.tlsroutes[k8s.NamespacedNameOf(obj)] = obj
-			return kc.routeTriggersRebuild(obj.Spec.ParentRefs)
+			return kc.routeTriggersRebuild(obj.Spec.ParentRefs), len(kc.tlsroutes)
+
 		case *gatewayapi_v1alpha2.GRPCRoute:
 			kc.grpcroutes[k8s.NamespacedNameOf(obj)] = obj
-			return kc.routeTriggersRebuild(obj.Spec.ParentRefs)
+			return kc.routeTriggersRebuild(obj.Spec.ParentRefs), len(kc.grpcroutes)
+
 		case *gatewayapi_v1beta1.ReferenceGrant:
 			kc.referencegrants[k8s.NamespacedNameOf(obj)] = obj
-			return true
+			return true, len(kc.referencegrants)
+
 		case *contour_api_v1alpha1.ExtensionService:
 			kc.extensions[k8s.NamespacedNameOf(obj)] = obj
-			return true
+			return true, len(kc.extensions)
+
 		case *contour_api_v1alpha1.ContourConfiguration:
-			return false
+			return false, 0
+
 		default:
 			// not an interesting object
 			kc.WithField("object", obj).Error("insert unknown object")
-			return false
+			return false, 0
 		}
 	}
 
-	if maybeInsert(obj) {
+	ok, count := maybeInsert(obj)
+	kind := k8s.KindOf(obj)
```

Comment:
if we keep it with this kind of implementation, what about instead of finding the count of resources in `maybeInsert`, we just take the kind and take the length of the map of the appropriate resource here (or even just iterate over all the kinds and set them all)

finding the number of resources to set the metric to above seems like it's easy to make a small mistake that is hard to catch, esp., in cases with special handling of resources that we don't necessarily want to add etc.

it would be good to add some testing to make sure we don't make any such mistakes

Reply:
@sunjayBhatia  thx,  @skriss WDYT? I will resend a new pr or update this one after we make a decision

Reply:
chatting with @skriss offline, I'm revising my position and in support of this impl:
- with my suggestion if any new resource kinds are added to the cache you will have to remember to update the component that sets the cache metrics, this impl forces you to b/c of the signature of `maybeInsert` etc.
- in the future we plan to attempt to utilize go generics more which should give us opportunities to refactor this

Reply:
when I get some time I'm planning on experimenting with making KubernetesCache use generics to simplify things and reduce type assertions/switches etc.